extensions/amp-ad/0.1/amp-ad-batch-manager.js

```diff
@@ -0,0 +1,150 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {user} from '../../../src/log';
+import {xhrFor} from '../../../src/xhr';
+
+/** @const {number} The number of milliseconds between checks to see if the response has arrived */
+const AD_BATCH_TICK_INTERVAL = 50;
+
+/** @const {number} The number of ticks to wait for a response from the ad server before timing out */
+const AD_BATCH_RETRIES = 100;
+
+/** @const {!string} Tag name for ad batch manager implementation. */
+export const TAG = 'amp-ad-batch-manager';
+
+/** @const {Object} A map of AmpAdBatchManager objects, one for each value of data-url on the page */
+let imageadBatchManagers = null;
+
+/** @const {Object} A map of arrays of imagead Elements, keyed by data-url */
+let imageadElements = null;
+/**
+  * Get a batch manager that will be used to batch together elements with the same data-url
+  * Note a side effect: the first ad with a given data-url to be processed will have its isBatchMaster_ variable set to true
+  * @param {Element} elem The element which is to be batched
+  * @param {string} url The URL identifying which batch
+  * @returns {AmpAdBatchManager}
+  */
+export function getBatchManager(elem, url) {
+  // If this is our first imagead, create a map of responses for each value of 'data-url'
+  // This allows ads from multiple ad servers on the same page
+  if (imageadBatchManagers === null) {
+    // Scan the page to get a list of all image ads on this page, indexed by url.
+    // For performance reasons, we only do the scan once, the first time that a batch manager is constructed.
+    imageadBatchManagers = {};
```

Comment:
Please use the [`map` helper](https://github.com/ampproject/amphtml/blob/master/src/types.js#L63-L69) to get a true map object.